lib/logstash/inputs/rabbitmq.rb

```diff
@@ -131,17 +139,87 @@ def consume!
         end
       end
 
+      public
       def stop
         super
         shutdown_consumer
         close_connection
       end
 
+      public
       def shutdown_consumer
         return unless @consumer
         @consumer.gracefully_shut_down
       end
 
+      private
+      def get_header_value(value)
+        # Two kinds of values require exceptional treatment:
+        #
+        # String values are instances of
+        # com.rabbitmq.client.impl.LongStringHelper.ByteArrayLongString
+        # and we don't want to propagate those.
+        #
+        # List values are java.util.ArrayList objects and we need to
+        # recurse into them to convert any nested strings values.
+        if value.respond_to? "toArray"
+          value.map{|item| get_header_value(item) }
+        elsif value.respond_to? "toString"
+          value.toString
+        else
+          value
+        end
+      end
+
+      private
+      def get_headers(metadata)
+        if not metadata.headers.nil?
+          Hash[metadata.headers.map {|k, v| [k, get_header_value(v)]}]
+        else
+          {}
+        end
+      end
+
+      private
+      def get_properties(metadata)
+        result = {}
+        # Map message properties to a lambda that's used to transform
+        # the value returned by the march_hare library into something
+        # suitable for a Logstash event.
+        #
+        # Technically the exchange, redeliver, and routing-key
+        # properties belong to the envelope and not the message but we
+        # ignore that distinction here. However, we extract the
+        # headers separately via get_headers even though the header
+        # table technically is a message property.
+        {
+          "app-id"           => lambda {|x| x },
+          "cluster-id"       => lambda {|x| x },
+          "consumer-tag"     => lambda {|x| x },
+          "content-encoding" => lambda {|x| x },
+          "content-type"     => lambda {|x| x },
+          "correlation-id"   => lambda {|x| x },
+          "delivery-mode"    => lambda {|x| x },
+          "exchange"         => lambda {|x| x },
+          "expiration"       => lambda {|x| x },
+          "message-id"       => lambda {|x| x },
+          "priority"         => lambda {|x| x },
+          "redeliver"        => lambda {|x| x },
+          "reply-to"         => lambda {|x| x },
+          "routing-key"      => lambda {|x| x },
+          # The AMQP 0.9.1 timestamp field only has second resolution
+          # so storing milliseconds serves no purpose and might give
+          # the incorrect impression of a higher resolution.
+          "timestamp"        => lambda {|x| x.getTime() / 1000 },
```

Comment:
Since we only have one field with a special case have considered reformatting this as:

```ruby
["app-id", "cluster-id", ...].reduce({}) do |acc,name|
  value = metadata.send(name.gsub("-", "_"))
  if value
    acc[name] = name != "timestamp" ? value : value.getTime / 1000
  end
  acc
end
```